### PR TITLE
Add DiscreteFrechetDistance

### PR DESCRIPTION
**Pull request by ssugiyama**
I‘ve just written an implementation of discrete fréchet distance according to http://www.kr.tuwien.ac.at/staff/eiter/et-archive/cdtr9464.pdf .  Review please.

### CODE REVIEW

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2009  Sandro Santilli <strk@keybit.net>
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: This belongs to you (and to this year).

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************
+ *
+ * Last port: algorithm/distance/DiscreteFrechetDistance.java 1.5 (JTS-1.10)
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: This isn't true, use "Last port: original work" (see src/operation/sharedpaths/SharedPathsOp.cpp for example)

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+ *
+ * An example where the default approximation is not close is:
+ * <pre>
+ *   A = LINESTRING (0 0, 100 0, 10 100, 10 100)
+ *   B = LINESTRING (0 100, 0 10, 80 10)
+ *
+ *   DHD(A, B) = 22.360679774997898
+ *   HD(A, B) ~= 47.8
+ * </pre>
+ */
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: Is this documentation correct ? As it looks like just the Hausdorff distance doc with a search & replace

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ * This algorithm is an approximation to the standard Frechet distance.
+ * Specifically,
+ * <pre>
+ *    for all geometries a, b:    DHD(a, b) <= HD(a, b)
+ * </pre>
+ * The approximation can be made as close as needed by densifying the
+ * input geometries.
+ * In the limit, this value will approach the true Frechet distance:
+ * <pre>
+ *    DHD(A, B, densifyFactor) -> HD(A, B) as densifyFactor -> 0.0
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: Should these functions be named DFD and FD ?

File: include/geos/algorithm/distance/Makefile.am [hunk 5]
```diff
@@ -1,13 +1,14 @@
 #
-# This file is part of project GEOS (http://trac.osgeo.org/geos/) 
+# This file is part of project GEOS (http://trac.osgeo.org/geos/)
```
Comment (strk) [inline include/geos/algorithm/distance/Makefile.am]: Please don't mix style and functional changes (I know, can be annoying until we make all files respect .editorconfig, sorry about that)

File: include/geos/algorithm/distance/PointPairDistance.h [hunk 6]
```diff
@@ -108,6 +108,10 @@ class PointPairDistance
 			initialize(p0, p1, dist);
 	}
 
+        bool getIsNull() {
+                return isNull;
+        }
+
```
Comment (strk) [inline include/geos/algorithm/distance/PointPairDistance.h]: The indent style of this block seems not to match the previous block, please check

File: src/algorithm/distance/DiscreteFrechetDistance.cpp [hunk 7]
```diff
@@ -0,0 +1,141 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2009  Sandro Santilli <strk@keybit.net>
```
Comment (strk) [inline src/algorithm/distance/DiscreteFrechetDistance.cpp]: same as above, copyright belongs to you and year to 2016

File: src/algorithm/distance/DiscreteFrechetDistance.cpp [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+ **********************************************************************/
+
+#include <geos/algorithm/distance/DiscreteFrechetDistance.h>
+#include <geos/geom/CoordinateSequence.h>
+
+//#include <geos/algorithm/CGAlgorithms.h>
+#include <geos/geom/Geometry.h>
+//#include <geos/geom/Polygon.h>
+//#include <geos/geom/GeometryCollection.h>
+//#include <geos/geom/Location.h>
```
Comment (strk) [inline src/algorithm/distance/DiscreteFrechetDistance.cpp]: you can drop the commented out includes

File: src/algorithm/distance/Makefile.am [hunk 9]
```diff
@@ -1,14 +1,15 @@
...
+AM_CPPFLAGS = -I$(top_srcdir)/include
 
 libdistance_la_SOURCES = \
     DiscreteHausdorffDistance.cpp \
-    DistanceToPoint.cpp 
+    DiscreteFrechetDistance.cpp \
+    DistanceToPoint.cpp
 
-libdistance_la_LIBADD = 
+libdistance_la_LIBADD =
```
Comment (strk) [inline src/algorithm/distance/Makefile.am]: please don't mix style and functional changes

**Comment by strk** [comment]
Impressive work, thank you ! I gave a first pass of review, will read again once all style changes are removed.

**Comment by strk** [comment]
Please consider also filing a trac ticket, so to schedule its inclusion:
https://trac.osgeo.org/geos/

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 10]
```diff
@@ -12,7 +12,10 @@
  *
  **********************************************************************
  *
- * Last port: algorithm/distance/DiscreteFrechetDistance.java 1.5 (JTS-1.10)
+ * Last port: original work
+ *
+ * Developed by Shinichi SUGIYAMA (shin.sugi@gmail.com)
+ ＊based on http://www.kr.tuwien.ac.at/staff/eiter/et-archive/cdtr9464.pdf
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: Is that an utf8 star I'm seeing there ? (thanks to my font that still lets me distinguish)

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 11]
```diff
@@ -71,30 +74,27 @@ namespace distance { // geos::algorithm::distance
...
  * <pre>
- *   A = LINESTRING (0 0, 100 0, 10 100, 10 100)
- *   B = LINESTRING (0 100, 0 10, 80 10)
+ *   A = LINESTRING (0 0, 100 0)
+ *   B = LINESTRING (0 0, 50 50, 100 0)
  *
- *   DHD(A, B) = 22.360679774997898
- *   HD(A, B) ~= 47.8
+ *   DFD(A, B) = 70.7106781186548
+ *   FD(A, B)  = 50.0
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: A section explaining how Frechet is different from Hausdorff may be useful.

File: include/geos/algorithm/distance/Makefile.am [hunk 12]
```diff
@@ -1,9 +1,9 @@
 #
-# This file is part of project GEOS (http://trac.osgeo.org/geos/)
+# This file is part of project GEOS (http://trac.osgeo.org/geos/) 
 #
-SUBDIRS =
+SUBDIRS = 
 
-EXTRA_DIST =
+EXTRA_DIST =
```
Comment (strk) [inline include/geos/algorithm/distance/Makefile.am]: These are still unrelated to the PR

File: include/geos/algorithm/distance/PointPairDistance.h [hunk 13]
```diff
@@ -108,7 +108,8 @@ class PointPairDistance
 			initialize(p0, p1, dist);
 	}
 
-        bool getIsNull() {
+        bool getIsNull()
+        {
```
Comment (strk) [inline include/geos/algorithm/distance/PointPairDistance.h]: This is also unrelated

**Comment by ssugiyama** [comment]
Thank you for the review and sorry for my lazy copy-and-paste. I've  just fixed the issues. Please review again.

File: include/geos/algorithm/distance/Makefile.am (hunk 12 above)
Comment (ssugiyama) [inline include/geos/algorithm/distance/Makefile.am]: > These are still unrelated to the PR
just restored the original trailing spaces to remove unrelated changes to the PR.

Comment (ssugiyama) [inline include/geos/algorithm/distance/Makefile.am]: Possibly, should I recreate the PR? ...

**Comment by strk** [comment]
On Tue, Nov 29, 2016 at 01:00:53AM -0800, Shinichi Sugiyama wrote:
> Possibly, should I recreate the PR? ...

It shouldn't be needed, just don't consider the first "approving"
review and only consider the second, we're almost there, I'd just
like to see the unrelated changes removed and an example of a case
in which Hausdorff and Frechet distance would give a different answer.

**Comment by ssugiyama** [comment]
Just added an example to show difference between HD and FD. Check it out.

**Comment by ssugiyama** [comment]
And, I think I've removed all the unrelated changes. Check them out, too.

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        PointPairDistance& getFrecheDistance(std::vector< std::vector<PointPairDistance> >& ca, size_t i, size_t j, const geom::CoordinateSequence& p, const geom::CoordinateSequence& q);
+
+	void compute(const geom::Geometry& discreteGeom,
+                     const geom::Geometry& geom);
+
+	const geom::Geometry& g0;
+
+	const geom::Geometry& g1;
+
+	PointPairDistance ptDist;
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: The 4 members above should be indented like the others

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+
+	/// Value of 0.0 indicates that no densification should take place
```
Comment (strk) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: Please indent the comment as the code below

File: include/geos/algorithm/distance/PointPairDistance.h [hunk 16]
```diff
@@ -108,6 +108,11 @@ class PointPairDistance
 			initialize(p0, p1, dist);
 	}
 
+        bool getIsNull()
+        {
+                return isNull;
+        }
+
```
Comment (strk) [inline include/geos/algorithm/distance/PointPairDistance.h]: Please use same indentation as the rest of the file

File: include/geos/algorithm/distance/PointPairDistance.h [hunk 17]
```diff
@@ -137,4 +142,3 @@ class PointPairDistance
 } // geos
 
 #endif // GEOS_ALGORITHM_DISTANCE_POINTPAIRDISTANCE_H
-
```
Comment (strk) [inline include/geos/algorithm/distance/PointPairDistance.h]: unrelated (style) change

File: tests/unit/algorithm/distance/DiscreteFrechetDistanceTest.cpp [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+		double distance = DiscreteFrechetDistance::distance(*g1,
+			*g2, densifyFactor);
+		double diff = std::fabs(distance-expectedDistance);
+		//std::cerr << "expectedDistance:" << expectedDistance << " actual distance:" << distance << std::endl;
+		ensure( diff <= TOLERANCE );
+	}
+
+	PrecisionModel pm;
+	GeometryFactory::unique_ptr gf;
+  geos::io::WKTReader reader;
```
Comment (strk) [inline tests/unit/algorithm/distance/DiscreteFrechetDistanceTest.cpp]: indent please

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h (hunk 14 above)
Comment (ssugiyama) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: fixed in 6b92c6a

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h (hunk 15 above)
Comment (ssugiyama) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: fixed in 6b92c6a

File: include/geos/algorithm/distance/PointPairDistance.h [hunk 19]
```diff
@@ -7,7 +7,7 @@
  *
  * This is free software; you can redistribute and/or modify it under
  * the terms of the GNU Lesser General Public Licence as published
- * by the Free Software Foundation. 
+ * by the Free Software Foundation.
```
Comment (ssugiyama) [inline include/geos/algorithm/distance/PointPairDistance.h]: fixed in 6b92c6a

File: tests/unit/Makefile.am [hunk 20]
```diff
@@ -1,5 +1,5 @@
 #
-# This file is part of project GEOS (http://trac.osgeo.org/geos/) 
+# This file is part of project GEOS (http://trac.osgeo.org/geos/)
```
Comment (ssugiyama) [inline tests/unit/Makefile.am]: fixed in 6b92c6a

File: include/geos/algorithm/distance/PointPairDistance.h (hunk 16 above)
Comment (ssugiyama) [inline include/geos/algorithm/distance/PointPairDistance.h]: fixed in 6b92c6a

File: include/geos/algorithm/distance/PointPairDistance.h (hunk 17 above)
Comment (ssugiyama) [inline include/geos/algorithm/distance/PointPairDistance.h]: fixed in 6b92c6a

File: tests/unit/algorithm/distance/DiscreteFrechetDistanceTest.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
+//
+// Test Suite for geos::algorithm::distance::DiscreteFrechetDistance
+// Ported from JTS junit/algorithm/distance/DiscreteFrechetDistanceTest.java rev. 1.2
```
Comment (ssugiyama) [inline tests/unit/algorithm/distance/DiscreteFrechetDistanceTest.cpp]: fixed in 6b92c6a

File: src/algorithm/distance/DiscreteFrechetDistance.cpp [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2016  Shinichi SUGIYAMA <shin.sugi@gmail.com>
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
```
Comment (ssugiyama) [inline src/algorithm/distance/DiscreteFrechetDistance.cpp]: fixed in 6b92c6a

**Comment by ssugiyama** [comment]
thank you so much! I've fixed the issues. check them out please.

**Comment by strk** [comment]
There are still style changes that don't belong here, but let's not get hung on that. I guess the style should be fixed once to make contributions easier. Please check with valgrind that the code doesn't leak or contain memory errors.

**Comment by ssugiyama** [comment]
> let's forget about style for now
> check with valgrind that the code doesn't leak or contain memory errors.

OK. Just running valgrind and found leaks. I've just fix them. Please check them out.

File: include/geos/algorithm/distance/DiscreteFrechetDistance.h [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+		return ptDist.getCoordinates();
+	}
+
+private:
+	geom::Coordinate getSegementAt(const geom::CoordinateSequence& seq, size_t index);
+
+	PointPairDistance& getFrecheDistance(std::vector< std::vector<PointPairDistance> >& ca, size_t i, size_t j, const geom::CoordinateSequence& p, const geom::CoordinateSequence& q);
+
+	void compute(const geom::Geometry& discreteGeom,
+		const geom::Geometry& geom);
```
Comment (ssugiyama) [inline include/geos/algorithm/distance/DiscreteFrechetDistance.h]: fixed in bac9a7d

File: src/algorithm/distance/DiscreteFrechetDistance.cpp [hunk 24]
```diff
@@ -147,6 +147,8 @@ DiscreteFrechetDistance::compute(
                 }
         }
         ptDist = getFrecheDistance(ca, pSize-1, qSize-1, *lp, *lq);
+	delete lp;
```
Comment (ssugiyama) [inline src/algorithm/distance/DiscreteFrechetDistance.cpp]: caused by mixture of tabs and spaces. I have replaced spaces with tabs in the files that use tab  indentation in bac9a7d

**Comment by ssugiyama** [comment]
I've noticed two mail addresses in my commits. I want unify them. So, can I reissue PR after all issues  are resolved?

**Comment by strk** [comment]
The official repository is an SVN one so GIT metainformation about
authorship would be lost anyway. I'll write your full name as part
of the commit log. Make sure email address is correct in the copyright
attribution line, and let me know if you want that email to be also
in the commit log.

**Comment by ssugiyama** [comment]
I see.

> Make sure email address is correct in the copyright
>  attribution line

it's correct.

> let me know if you want that email to be also 
> in the commit log

I do. 

Thanks!

**Comment by strk** [comment]
Please fix errors returned by:
```
$ git diff --check svn-trunk
include/geos/algorithm/distance/DiscreteFrechetDistance.h:117: space
before tab in indent.
+                               const geom::Geometry& g1)
src/algorithm/distance/DiscreteFrechetDistance.cpp:130: trailing
whitespace.
+       if (densifyFrac > 0) 
src/algorithm/distance/DiscreteFrechetDistance.cpp:136: trailing
whitespace.
+       else 
src/algorithm/distance/DiscreteFrechetDistance.cpp:142: trailing
whitespace.
+       for (size_t i = 0; i < pSize; i++) 
src/algorithm/distance/DiscreteFrechetDistance.cpp:144: trailing
whitespace.
+               for (size_t j = 0; j < qSize; j++) 
```

**Comment by ssugiyama** [comment]
fixed the errors in 273b8fa

    $ git diff --check svn-trunk
    (empty)

**Comment by strk** [comment]
Thanks.
Next: could you add a test under unit/capi/, for the new C-API interface ?
Unfortunately a test is also lacking for HausdorffDistance so you
won't be able to copy&paste, but you could look at GEOSDistanceTest.cpp
for inspiration.

Since you're at it, if you want to also do the HausdorffDistance one
you're welcome :)

**Comment by ssugiyama** [comment]
Added both test cases!

**Comment by strk** [comment]
Thanks, last bit: squash and rebase to svn-trunk branch, and make sure `git show --check` is still good, then we go :)

**Comment by ssugiyama** [comment]
just tried in my local repo as below:

    % git checkout svn-trunk
    % git pull origin svn-trunk
    % git checkout discrete-frechet-distance
    % git rebase -i svn-trunk
    (squash commits in editor)
    % git diff --check svn-trunk
    (empty output)

Is this OK?

**Comment by strk** [comment]
Yes, sounds good, please add a reference to your own name/email, a reference to a trac ticket (did you file it yet?) on https://trac.osgeo.org/geos/ in the now-squashed commit log and force-push to the branch

**Comment by strk** [comment]
Found the trac ticket: https://trac.osgeo.org/geos/ticket/797
Reference can just be `Closes #797`

**Comment by ssugiyama** [comment]
thank you! I've just force-pushed to the branch.

**Comment by strk** [comment]
Uhm, github still reports your branch to be out-of-date with the base branch, what's going on ?
Did you rebase against your own vision of `svn-trunk` ? That is, is your `origin` remote your fork rather than the official mirror ?

**Comment by ssugiyama** [comment]
sorry. I should have pulled from upstream not my forked repo. I've force-pushed again.

**Comment by strk** [comment]
r4314 = 596e4fefa4561afeab9226eb8682ecd6136acc05 (refs/remotes/trunk)
